src/toil/batchSystems/singleMachine.py

```diff
@@ -824,20 +936,93 @@ def acquisitionOf(self, amount):
-            current demands, as well as the amount requested and amount actually available.
+class ResourceSet:
+    """
+    Represents a collection of distinct resources (such as accelerators).
+
+    Subsets can be acquired immediately or with a timeout, and released.
+
+    Provides a context manager to do something with a set of of resources
+    acquired.
+    """
+    def __init__(self, initial_value: Set[int], resource_type: str, timeout: float = 5):
+        super().__init__()
+        # We use this condition to signal everyone whenever some resource is released.
+        # We use its associated lock to guard value.
+        self.condition = Condition()
+        # This records what resources are available right now.
+        self.value = initial_value
+        self.resource_type = resource_type
+        self.timeout = timeout
+
+    def acquireNow(self, subset: Set[int]):
+        """
+        Reserve the given amount of the given resource.
+
+        Returns True if successful and False if this is not possible immediately.
+        """
+
+        with self.condition:
+            if subset > self.value:
+                return False
+            self.value -= subset
+            self.__validate()
+            return True
+
+    def acquire(self, subset: Set[int]):
+        """
+        Reserve the given amount of the given resource.
+
+        Raises AcquisitionTimeoutException if this is not possible in under
+        self.timeout time.
+        """
+        with self.condition:
+            startTime = time.time()
+            while subset > self.value:
+                if time.time() - startTime >= self.timeout:
+                    # This means the thread timed out waiting for the resource.
+                    raise AcquisitionTimeoutException(resource=self.resource_type,
+                                                      requested=subset, available=self.value)
+                # Allow self.timeout seconds to get the resource, else quit
+                # through the above if condition. This wait + timeout is the
+                # last thing in the loop such that a request that takes longer
+                # than self.timeout due to multiple wakes under the threshold
+                # are still honored.
+                self.condition.wait(timeout=self.timeout)
+            self.value -= subset
+            self.__validate()
+
+    def release(self, subset: Set[int]):
+        with self.condition:
+            self.value |= subset
+            self.__validate()
+            self.condition.notify_all()
+
+    def get_free_snapshot(self) -> Set[int]:
+        """
+        Get a snapshot of what items are free right now.
+        May be stale as soon as you get it, but you will need some kind of hint
+        to try and do an acquire.
+        """
+        return set(self.value)
+
+    def __validate(self):
+        # Nothing to do!
+        pass
```

Comment:
Should `__validate()` and its use in this class be removed?  `__validate()` only exists for this class and won't be overwritten by inheritance due to the nature of how double underscores are treated in python.

Reply:
That makes sense. I kind of was hoping to re-combine the classes somehow, but there's no point in having the method in this copy and removing it shouldn't make that much harder later.